### PR TITLE
Stop request to Brave News CDN when Brave News is not enabled / opted in

### PR DESCRIPTION
**Pull request by fallaciousreasoning**
Resolves https://github.com/brave/brave-browser/issues/28268

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

1. In a new browser profile, open the NTP. No request should be made to the Brave News CDN
2. Opt in to Brave News - a request should be made to the CDN and Brave News should be enabled.

### CODE REVIEW

File: components/brave_today/common/pref_names.cc [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/brave_today/common/pref_names.h"
+
+#include "components/prefs/pref_service.h"
+
+namespace brave_news {
+bool GetIsEnabled(PrefService* prefs) {
+  bool should_show = prefs->GetBoolean(prefs::kNewTabPageShowToday);
```
Comment (sangwoo108) [inline components/brave_today/common/pref_names.cc:12]: I think it'd be safer to redirect |prefs| to the original |prefs|. Previously, BrowserNewsControllerFactory::GetForContext() was doing the redirection but now that we have this standalone function, we might need that. 

```suggestion
  prefs = prefs->GetOriginalProfile();
  bool should_show = prefs->GetBoolean(prefs::kNewTabPageShowToday);
```

There could be more options for us here, but I"m not sure what is the best way.

* Add DCHECK() to prefs is not incognito?
* Or, GetIngognitoPersistentPrefsAllowlist? - this could be the most recommended way https://bravesoftware.slack.com/archives/C7VLGSR55/p1670000442736269

File: components/brave_today/browser/channels_controller.cc [hunk 2]
```diff
@@ -54,9 +55,13 @@ void ChannelsController::SetChannelSubscribedPref(PrefService* prefs,
...
     PrefService* prefs,
     PublishersController* publishers_controller)
-    : prefs_(prefs), publishers_controller_(publishers_controller) {}
+    : prefs_(prefs), publishers_controller_(publishers_controller) {
+  publishers_controller_->AddObserver(this);
+}
 
-ChannelsController::~ChannelsController() = default;
+ChannelsController::~ChannelsController() {
+  publishers_controller_->RemoveObserver(this);
```
Comment (sangwoo108) [inline components/brave_today/browser/channels_controller.cc:63]: nit: Using base::ScopedObservation would be better

**Review by sangwoo108 — APPROVED** [review]
LGTM with a few nits! Nice work

File: components/brave_today/common/pref_names.cc (hunk 1 above)
Comment (petemill) [reply components/brave_today/common/pref_names.cc:12]: Maybe add `DCHECK`? My initial concern with `GetOriginalProfile` is that if some UI in the browser calls this function to see if they should enable or show Brave News and that UI is also in the private window, then the feature will be active for the private profile, when it should not? Though I guess that only relies on the defaults being `false` which may always be that way.

Comment (fallaciousreasoning) [reply components/brave_today/common/pref_names.cc:12]: Maybe we should just return false in incognito windows?

```suggest
prefs = prefs.GetOriginalProfile();
...
return !Incognito() && should_show && opted_in
```

Comment (sangwoo108) [reply components/brave_today/common/pref_names.cc:12]: Yeah, if brave news should be disabled on InCognito(), that would be better. In addition, if we should disable it on guest profiles too, then !IsOffTheRecord() would be even better.

And in favor of that, I think we should remove `BraveNewsControllerFactory::GetBrowserContextToUse` override at some point. This could misleading as if we could support Brave News on incognito.

File: components/brave_today/browser/channels_controller.cc (hunk 2 above)
Comment (fallaciousreasoning) [reply components/brave_today/browser/channels_controller.cc:63]: I always forget it exists :laughing: :+1:

File: components/brave_today/common/pref_names.cc (hunk 1 above)
Comment (fallaciousreasoning) [reply components/brave_today/common/pref_names.cc:12]: Hey, so is there an easy way to tell if you're `OffTheRecord` from components? There isn't a `GetOriginalProfile` method or any flags on `PrefService`.

Comment (sangwoo108) [reply components/brave_today/common/pref_names.cc:12]: Oh, that's a good point. Could you try passing BrowserContext here and using BrowserContext::IsOffTheRecord?

**Comment by fallaciousreasoning** [comment]
@goodov for `DEPS` file changes

File: components/brave_today/common/pref_names.cc (hunk 1 above)
Comment (goodov) [reply components/brave_today/common/pref_names.cc:17]: I assume we're fixing two issues in this PR:
1. BraveNews requests on a clean profile (a check added into `PublishersController::GetOrFetchPublishers` solves this).
2. BraveNews requests in incognito if it's enabled in the main profile (a `BrowserContext` passed everywhere solves this).

> And in favor of that, I think we should remove `BraveNewsControllerFactory::GetBrowserContextToUse` override at some point. This could misleading as if we could support Brave News on incognito.

I don't understand why are we passing `BrowserContext` to `GetIsEnabled` to disable `BraveNews` interaction in incognito, instead of just not creating `BraveNews`? Is there something that blocks us from doing so? If we can do this in this PR and keep things simple (nothing explodes, we don't need refactoring), then do it right here. If that's not the case, then we can fix (1) in this PR and do a separate PR to disable BraveNews in incognito entirely.

wdyt?

Comment (sangwoo108) [reply components/brave_today/common/pref_names.cc:17]: Yup, I was wondering, removing `BraveNewsControllerFactory::GetBrowserContextToUse` could cause some code to access invalid nullptr. If scanning call sites and adding validity checks are not that heavy burden, removing it from this PR would be ideal.

Comment (fallaciousreasoning) [reply components/brave_today/common/pref_names.cc:13]: Seems to have been pretty painless to remove, to be honest. Thanks @goodov, this is definitely nicer.

**Review by petemill — APPROVED** [review]

File: browser/brave_news/brave_news_controller_factory.cc [hunk 3]
```diff
@@ -60,7 +60,7 @@ BraveNewsControllerFactory::~BraveNewsControllerFactory() = default;
 
 KeyedService* BraveNewsControllerFactory::BuildServiceInstanceFor(
     content::BrowserContext* context) const {
-  if (!brave::IsRegularProfile(context)) {
+  if (!brave::IsRegularProfile(context) || context->IsOffTheRecord()) {
```
Comment (goodov) [inline browser/brave_news/brave_news_controller_factory.cc:63]: you shouldn't need this check as `GetBrowserContextToUse` default version is to NOT create a service in Incognito.

File: components/brave_today/browser/DEPS [hunk 4]
```diff
@@ -1,6 +1,7 @@
 include_rules = [
+  "+content/public/browser",
```
Comment (goodov) [inline components/brave_today/browser/DEPS:2]: please remove all unrelated changes from this PR.

**Comment by fallaciousreasoning** [comment]
@goodov just checking you don't have anything more to add?

File: components/brave_today/common/pref_names.cc [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/brave_today/common/pref_names.h"
+
+#include "components/prefs/pref_service.h"
+
+namespace brave_news {
+bool GetIsEnabled(PrefService* prefs) {
```
Comment (goodov) [inline components/brave_today/common/pref_names.cc:11]: this function should be in `brave_today/browser`.

Comment (fallaciousreasoning) [reply components/brave_today/common/pref_names.cc]: Changed, but why does it belong in `browser`? Is it because it depends on `components/prefs`?

**Review by goodov — APPROVED** [review]
